# Project 4 - Hackathon (Predicting Income under Team Features Constraint)

* Choose algorithm and choice of samples
* Limited to a maximum of 20 features

### This exercise is to build the best model possible under those constraints. 

### The task is to predict if a person's income is in excess of 50,000 dollars given certain profile information, and more specifically to generate predicted probabilities of income being above 50,000 dollars for each row in the test set. The output will be a .csv file with a single column of the probability with 'wage' as a header. The file is to be submitted by the end of the day.

### This section is the Modeling  portion. 

Perform the following model and evaluate the results

- Logistic Regresion 
- Knn Regression


The data set is the 'train_clean_data.cvs' file.
 

In [1]:
# Import some important stuff
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from scipy import stats

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score



In [2]:
# This funtion was created by Magnus Bigelow
# Function to calculate and display classification metrics, works for bernoulli y
def class_metrics(model, X, y):
    # Generate predictions
    preds = model.predict(X)
    # Get confusion matrix and unravel
    tn, fp, fn, tp = confusion_matrix(y,preds).ravel()
    # Accuracy
    print(f'Accuracy: {round((tp+tn)/len(y),3)}')
    # Sensitivity
    print(f'Sensitivity: {round(tp/(tp+fn),3)}')
    # Specificity
    print(f'Specificity: {round(tn/(tn+fp),3)}')
    # Precision
    print(f'Precision: {round(tp/(tp+fp),3)}')

In [3]:
# This funtion was created by Magnus Bigelow
# Function to print train and test F1 score
def f1(model, X_train, y_train, X_test, y_test):
    y_train_p = model.predict(X_train)
    y_test_p = model.predict(X_test)
    f_train = f1_score(y_train,y_train_p)
    f_test = f1_score(y_test,y_test_p)
    print(f'Train F1: {round(f_train,3)}')
    print(f'Test F1: {round(f_test,3)}')

In [4]:
np.random.seed(33) 

In [5]:
df_data_clean = pd.read_csv('./data/clean_train.csv')

In [6]:
df_data_clean.shape

(32561, 20)

In [7]:
df_data_clean.head().T

0         1         2         3  \
age                               39.00     50.00     38.00     53.00   
education-num                     13.00     13.00      9.00      7.00   
sex                                1.00      1.00      1.00      1.00   
capital-gain                    2174.00      0.00      0.00      0.00   
capital-loss                       0.00      0.00      0.00      0.00   
hours-per-week                    40.00     13.00     40.00     40.00   
wage                               0.00      0.00      0.00      0.00   
marital_status_num                 0.00      1.00      0.00      1.00   
occupation_com_House_Services      0.00      0.00      1.00      1.00   
occupation_com_Other               1.00      0.00      0.00      0.00   
occupation_com_Professional        0.00      1.00      0.00      0.00   
occupation_com_Specialty           0.00      0.00      0.00      0.00   
occupation_com_Tech/sales          0.00      0.00      0.00      0.00   
workclass_com_ Government          1.00      0.00      0.00      0.00   
workclass_com_ Other               0.00      0.00      0.00      0.00   
workclass_com_ Private             0.00      0.00      1.00      1.00   
workclass_com_ Self-employed       0.00      1.00      0.00      0.00   
cap_gain_binary                    1.00      0.00      0.00      0.00   
cap_loss_binary                    0.00      0.00      0.00      0.00   
gdp_pc                         41524.09  41524.09  41524.09  41524.09   

                                       4  
age                               28.000  
education-num                     13.000  
sex                                1.000  
capital-gain                       0.000  
capital-loss                       0.000  
hours-per-week                    40.000  
wage                               0.000  
marital_status_num                 1.000  
occupation_com_House_Services      0.000  
occupation_com_Other               0.000  
occupation_com_Professional        1.000  
occupation_com_Specialty           0.000  
occupation_com_Tech/sales          0.000  
workclass_com_ Government          0.000  
workclass_com_ Other               0.000  
workclass_com_ Private             1.000  
workclass_com_ Self-employed       0.000  
cap_gain_binary                    0.000  
cap_loss_binary                    0.000  
gdp_pc                         12492.097

In [8]:
df_data_clean.columns

Index(['age', 'education-num', 'sex', 'capital-gain', 'capital-loss',
       'hours-per-week', 'wage', 'marital_status_num',
       'occupation_com_House_Services', 'occupation_com_Other',
       'occupation_com_Professional', 'occupation_com_Specialty',
       'occupation_com_Tech/sales', 'workclass_com_ Government',
       'workclass_com_ Other', 'workclass_com_ Private',
       'workclass_com_ Self-employed', 'cap_gain_binary', 'cap_loss_binary',
       'gdp_pc'],
      dtype='object')

In [9]:
list = ['age', 'education-num', 'sex', 'hours-per-week', 'marital_status_num',
       'occupation_com_House_Services',
       'occupation_com_Professional', 'occupation_com_Specialty',
       'occupation_com_Tech/sales', 'workclass_com_ Government', 'workclass_com_ Private',
       'workclass_com_ Self-employed', 'cap_gain_binary', 'cap_loss_binary', 'gdp_pc']

In [10]:
# Routine to create a combinations of features combination to try out in a Logistic Regression
# set up list data
df_data = pd.DataFrame()
data = {}

ndx = 0
ndx_1 = 0


while ndx_1 < len(list):
#    data = {'name':[list[ndx_1]]}
#    df_data = df_data.append(data, ignore_index = True)
    ndx += 1
    ndx_2 = ndx_1 + 1
    while ndx_2 < len(list):
        data = {'name':[list[ndx_1],list[ndx_2]]}
        df_data = df_data.append(data, ignore_index = True)
        ndx_3 = ndx_2 + 1
        while ndx_3 < len(list):
            data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3]]}
            df_data = df_data.append(data, ignore_index = True)
            ndx_4 = ndx_3 + 1
            while ndx_4 < len(list):
                data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4]]}
                df_data = df_data.append(data, ignore_index = True)
                ndx_5 = ndx_4 + 1
                while ndx_5 < len(list):
                    data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5]]}
                    df_data = df_data.append(data, ignore_index = True)
                    ndx_6 = ndx_5 + 1
                    while ndx_6 < len(list):
                        data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6]]}
                        df_data = df_data.append(data, ignore_index = True)
                        ndx_7 = ndx_6 + 1
                        while ndx_7 < len(list):
                            data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7]]}
                            df_data = df_data.append(data, ignore_index = True)
                            ndx_8 = ndx_7 + 1
                            while ndx_8 < len(list):
                                data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8]]}
                                df_data = df_data.append(data, ignore_index = True)
                                ndx_9 = ndx_8 + 1
                                while ndx_9 < len(list):
                                    data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9]]}
                                    df_data = df_data.append(data, ignore_index = True)
                                    ndx_10 = ndx_9 + 1
                                    while ndx_10 < len(list):
                                        data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9],list[ndx_10]]}
                                        df_data = df_data.append(data, ignore_index = True)
                                        ndx_11 = ndx_10 + 1
                                        while ndx_11 < len(list):
                                            data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9],list[ndx_10],list[ndx_11]]}
                                            df_data = df_data.append(data, ignore_index = True)
                                            ndx_12 = ndx_11 + 1
                                            while ndx_12 < len(list):
                                                data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9],list[ndx_10]]}
                                                df_data = df_data.append(data, ignore_index = True)
                                                ndx_13 = ndx_12 + 1
                                                while ndx_13 < len(list):
                                                    data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9],list[ndx_10],list[ndx_11]]}
                                                    df_data = df_data.append(data, ignore_index = True)
                                                    ndx_14 = ndx_13 + 1
                                                    while ndx_14 < len(list):
                                                        data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9],list[ndx_10]]}
                                                        df_data = df_data.append(data, ignore_index = True)
                                                        ndx_15 = ndx_14 + 1
                                                        while ndx_15 < len(list):
                                                            data = {'name':[list[ndx_1],list[ndx_2],list[ndx_3],list[ndx_4],list[ndx_5],list[ndx_6],list[ndx_7],list[ndx_8],list[ndx_9],list[ndx_10],list[ndx_11]]}
                                                            df_data = df_data.append(data, ignore_index = True)
                                                            ndx_15 += 1
                                                        ndx_14 += 1
                                                    ndx_13 += 1      
                                                ndx_12 += 1                
                                            ndx_11 += 1
                                        ndx_10 += 1
                                    ndx_9 += 1      
                                ndx_8 += 1                
                            ndx_7 += 1
                        ndx_6 += 1
                    ndx_5 += 1      
                ndx_4 += 1                
            ndx_3 += 1
        ndx_2 += 1          
    ndx_1 += 1

In [11]:
df_results_LogReg = pd.DataFrame(columns=['train_score', 'test_score', 'diff', 'train_f1_score', 'test_f1_score', 
       'Accuracy', 'Sensitivity', 'Specificity', 'Precision', 'age', 'education-num', 'sex', 'hours-per-week',
       'marital_status_num', 'occupation_com_House_Services', 'occupation_com_Professional', 'occupation_com_Specialty',
       'occupation_com_Tech/sales', 'workclass_com_ Government', 'workclass_com_ Private',
       'workclass_com_ Self-employed', 'cap_gain_binary', 'cap_loss_binary', 'gdp_pc'])

In [ ]:
# This routine will iterate through the list of features combination to get information on 
#     the training and testing score along with the coeficient with each feature used.
ndx = 0
y = df_data_clean['wage'] 
logreg = LogisticRegression(solver='liblinear')

while ndx < len(df_data):
    X = df_data_clean[df_data['name'][ndx]] 
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.33, random_state=33)
    logreg.fit(X_train, y_train)
    
    temp_dict = {}

    temp_dict = {'train_score': round(logreg.score(X_train, y_train), 4),  
                 'test_score': round(logreg.score(X_test, y_test), 4),
                 'diff': round(logreg.score(X_train, y_train) - logreg.score(X_test, y_test), 4),
                 'train_f1_score': round(f1_score(y_train,logreg.predict(X_train)), 4),  
                 'test_f1_score': round(f1_score(y_test,logreg.predict(X_test)), 4)}
    
    tn, fp, fn, tp = confusion_matrix(y_test, logreg.predict(X_test)).ravel()
                 
    temp_dict.update({'Accuracy': round((tp+tn)/len(y_test), 4),  
                 'Sensitivity': round(tp/(tp+fn), 4), 
                 'Specificity': round(tn/(tn+fp), 4),     
                 'Precision': round(tp/(tp+fp), 4)})     
    
    ndx_x = 0
    while ndx_x < len(X.columns):
        temp_dict.update({X.columns.values[ndx_x] : np.round(logreg.coef_[0][ndx_x], 4)})
        ndx_x += 1
    
    df_results_LogReg = df_results_LogReg.append(temp_dict, ignore_index=True)
    
    ndx += 1 



In [16]:
df_results_LogReg.sort_values(by='train_score', ascending = False).head()

train_score  test_score    diff  train_f1_score  test_f1_score  \
681        0.8426      0.8339  0.0087          0.6389         0.6175   
4752       0.8425      0.8343  0.0082          0.6381         0.6185   
4776       0.8425      0.8340  0.0086          0.6386         0.6177   
4745       0.8424      0.8344  0.0080          0.6382         0.6185   
4649       0.8424      0.8346  0.0077          0.6377         0.6181   

      Accuracy  Sensitivity  Specificity  Precision     age  ...  \
681     0.8339       0.5557       0.9224     0.6948  0.0287  ...   
4752    0.8343       0.5569       0.9225     0.6956  0.0285  ...   
4776    0.8340       0.5557       0.9225     0.6951  0.0280  ...   
4745    0.8344       0.5565       0.9227     0.6961  0.0285  ...   
4649    0.8346       0.5546       0.9237     0.6981  0.0280  ...   

      occupation_com_House_Services  occupation_com_Professional  \
681                             NaN                       1.0085   
4752                            NaN                       0.9901   
4776                            NaN                       1.0143   
4745                            NaN                       0.9854   
4649                            NaN                       1.1047   

      occupation_com_Specialty  occupation_com_Tech/sales  \
681                        NaN                     0.7019   
4752                       NaN                     0.6830   
4776                       NaN                     0.7020   
4745                       NaN                     0.6787   
4649                    0.1634                     0.7984   

      workclass_com_ Government  workclass_com_ Private  \
681                         NaN                  0.1340   
4752                     0.2877                  0.3466   
4776                        NaN                  0.1199   
4745                     0.3348                  0.3935   
4649                        NaN                  0.0934   

      workclass_com_ Self-employed  cap_gain_binary  cap_loss_binary  gdp_pc  
681                        -0.2111           1.6635           1.1940     NaN  
4752                           NaN           1.6629           1.1896     NaN  
4776                       -0.2162           1.6598           1.1907     NaN  
4745                        0.0583           1.6634           1.1900     NaN  
4649                       -0.2507           1.6600           1.1886     NaN  

[5 rows x 24 columns]

In [17]:
# save results
df_results_LogReg.to_csv('./data/logreg_results.csv', index=False)

From the results, the best training score was 0.8426 with a testing score of 0.8339. The f1_test_score was 0.6175.
Accuracy: 0.8339, Senitivity: 0.557, Specificity: 0.992, Precision: 0.6948.

The best features found was age, education-num, sex, marital_tatus_num, occupation_com_Professional, occupation_com_tech/sales, workclass_com_Government, workclass_com_Private, cap_gain_binary, and cap_loss_binary

10 features found.

# Knn model

In [18]:
X = df_data_clean[['age', 'education-num', 'sex',
       'hours-per-week', 'marital_status_num',
       'occupation_com_House_Services',
       'occupation_com_Professional', 'occupation_com_Specialty',
       'occupation_com_Tech/sales', 'workclass_com_ Government', 'workclass_com_ Private',
       'workclass_com_ Self-employed', 'cap_gain_binary', 'cap_loss_binary',
       'gdp_pc']]

y = df_data_clean['wage'] 

In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.33, random_state=33)

In [20]:
# will need to scale the data
ss = StandardScaler()
X_train_sc = ss.fit_transform(X_train)
X_test_sc = ss.transform(X_test)

In [21]:
# Instantiate our model (use default parameters).
knn = KNeighborsClassifier()

In [22]:
# Fit our model to our data.
knn.fit(X_train_sc, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [23]:
# Get the training score
knn.score(X_train_sc, y_train)

0.8714187485674995

In [24]:
# Get the testing score
knn.score(X_test_sc, y_test)

0.8203052298529685

In [25]:
f1(knn, X_train_sc, y_train, X_test_sc, y_test)

Train F1: 0.72
Test F1: 0.607


In [26]:
class_metrics(knn, X_test_sc, y_test)

Accuracy: 0.82
Sensitivity: 0.575
Specificity: 0.898
Precision: 0.643


From our analysis from all the modeling, we determined that the best model to use was the Knn because it gave us a similiar score, with the best sensitivity. We were looking for a good balance model.

The next section is to run the Knn through a few hyperparameters, to get the best values.



# Knn Through GridSearch



In [27]:
knn_params = {'n_neighbors' : [3, 5, 15, 25],
              'p':[1,2,3]}


In [28]:
knn_gridsearch = GridSearchCV(KNeighborsClassifier(), 
                              knn_params, 
                              cv=5, 
                              verbose=1)

In [29]:
knn_gridsearch.fit(X_train_sc, y_train);

Fitting 5 folds for each of 12 candidates, totalling 60 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  60 out of  60 | elapsed:  1.7min finished


In [30]:
knn_gridsearch.best_params_

{'n_neighbors': 25, 'p': 1}

In [31]:
# Get the training score
knn_gridsearch.score(X_train_sc, y_train)

0.849690579876232

In [32]:
# Get the testing score
knn_gridsearch.score(X_test_sc, y_test)

0.8368695328494323

In [33]:
f1(knn_gridsearch, X_train_sc, y_train, X_test_sc, y_test)

Train F1: 0.663
Test F1: 0.635


In [34]:
class_metrics(knn_gridsearch, X_test_sc, y_test)

Accuracy: 0.837
Sensitivity: 0.589
Specificity: 0.916
Precision: 0.69


### We will set the n_neighbor to 25 and p to 1  when setting up our testing data

code will be in the test_workbook